### PR TITLE
Worker lib

### PR DESCRIPTION
**Pull request by alexmeli100**

### CODE REVIEW

File: supervisor/process.go [hunk 1]
```diff
@@ -0,0 +1,411 @@
...
+		return "starting"
+	case Running:
+		return "running"
+	case Stopping:
+		return "stopping"
+	case Finished:
+		return "finished"
+	case Fatal:
+		return "fatal"
+	default:
```
Comment (r0mant) [inline supervisor/process.go]: Feels like Interrupted is missing from here.

File: supervisor/process.go [hunk 2]
```diff
@@ -0,0 +1,411 @@
...
+	return process
+
+}
+
+// setState sets the state of a process
+func (p *Process) setState(state ProcessState) {
+	// cannot the set the state of a finished process
+	if p.state == Finished || p.state == state {
+		return
+	}
```
Comment (r0mant) [inline supervisor/process.go]: Shouldn't this be under lock as well?

File: supervisor/process.go [hunk 3]
```diff
@@ -0,0 +1,411 @@
...
+
+	p.stateLock.Lock()
+	defer p.stateLock.Unlock()
+
+	// the next state can only be Starting if the state is Initial
+	if p.state == Initial {
+		p.state = Starting
```
Comment (r0mant) [inline supervisor/process.go]: This implicit logic may be unexpected for the caller, why not just set the state to whatever is passed in (and have the caller pass proper state)?

File: supervisor/process.go [hunk 4]
```diff
@@ -0,0 +1,411 @@
...
+		p.state = state
+	}
+
+	p.status.State = p.state
+}
+
+// IsInitialState is true if process state is Initial
+func (p *Process) IsInitialState() bool {
+	p.stateLock.Lock()
+	defer p.stateLock.Unlock()
```
Comment (r0mant) [inline supervisor/process.go]: Could switch to r/w mutex instead and use RLock for operations that don't change the state.

File: supervisor/process.go [hunk 5]
```diff
@@ -0,0 +1,411 @@
...
+
+			}
+		}
+	}
+
+	// set final status of the process
+	p.status.Error = err
+	p.status.ExitCode = exitCode
+	p.status.StopTime = now.UnixNano()
+	p.setState(Finished)
```
Comment (r0mant) [inline supervisor/process.go]: This will override Interrupted status if it was set above?

File: supervisor/process.go [hunk 6]
```diff
@@ -0,0 +1,411 @@
...
+}
+
+// GetStatus returns the current status of the process
+func (p *Process) GetStatus() *Status {
+	p.Lock()
+	defer p.Unlock()
+
+	return &p.status
```
Comment (r0mant) [inline supervisor/process.go]: Return a copy by value instead.

File: supervisor/supervisor.go [hunk 7]
```diff
@@ -0,0 +1,314 @@
...
+
+	st := job.GetStatus()
+
+	return &JobStatus{id, st}, nil
+}
+
+// GetOutput returns the line buffer for the job
+func (s *Supervisor) GetOutput(id string) ([]LogOutput, error) {
+	s.access.Lock()
+	s.access.Unlock()
```
Comment (r0mant) [inline supervisor/supervisor.go]: defer?

File: supervisor/process.go [hunk 8]
```diff
@@ -0,0 +1,411 @@
...
+		newLineOffset := bytes.IndexByte(b[firstChar:], '\n')
+
+		if newLineOffset < 0 {
+			break // no newlines in stream
+		}
+
+		lastCharOffset := firstChar + newLineOffset
+
+		// check for carriage return and strip it if present
+		if newLineOffset > 0 && b[lastCharOffset-1] == '\r' {
```
Comment (r0mant) [inline supervisor/process.go]: Why is this needed?

File: supervisor/process.go [hunk 9]
```diff
@@ -0,0 +1,411 @@
...
+
+		var line string
+
+		// preprend any lines if buffer is not empty
+		if s.lastChar > 0 {
+			line = string(s.buf[0:s.lastChar])
+			s.lastChar = 0
+		}
+
+		line += string(b[firstChar:lastCharOffset])
```
Comment (r0mant) [inline supervisor/process.go]: All this logic is pretty complicated and looks pretty fragile TBH. Would things simplify if you didn't try to split the output by lines and just stream everything byte by byte?

File: supervisor/process.go [hunk 10]
```diff
@@ -0,0 +1,411 @@
...
+
+		s.streamChan <- line
```
Comment (r0mant) [inline supervisor/process.go]: Since you're pushing to the channel on write, how will the clients connecting to the job receive prior logs if the job doesn't produce output anymore?

File: supervisor/supervisor.go [hunk 11]
```diff
@@ -0,0 +1,314 @@
...
+		return nil, ErrJobNotFound
+	}
+
+	if err := job.Stop(true); err != nil {
+		s.logger.Error("Cannot stop process", zap.String("id", id))
+		return nil, errors.Wrap(err, "cannot stop process")
+	}
+
+	// Wait for process to exit completely
+	<-job.Done()
```
Comment (r0mant) [inline supervisor/supervisor.go]: When Stop is called over API, this will hold the client on the wire which is probably not a great UX.

File: supervisor/process.go (hunk 1 above)
Comment (alexmeli100) [reply supervisor/process.go]: Thanks. This has been fixed

File: supervisor/supervisor.go (hunk 7 above)
Comment (alexmeli100) [reply supervisor/supervisor.go]: This been fixed. Don't know how I missed this

File: supervisor/process.go (hunk 3 above)
Comment (alexmeli100) [reply supervisor/process.go]: Yeah I thought some basic validation could help the caller avoid mistakes. This logic has been removed

File: supervisor/process.go (hunk 2 above)
Comment (alexmeli100) [reply supervisor/process.go]: This has been fixed as well

File: supervisor/process.go (hunk 6 above)
Comment (alexmeli100) [reply supervisor/process.go]: This has been fixed. Thanks

File: supervisor/process.go (hunk 5 above)
Comment (alexmeli100) [reply supervisor/process.go]: Yes it overrides the Interrupted status

File: supervisor/process.go (hunk 4 above)
Comment (alexmeli100) [reply supervisor/process.go]: This has been fixed. Thanks

File: supervisor/supervisor.go (hunk 11 above)
Comment (alexmeli100) [reply supervisor/supervisor.go]: Thanks. This has been fixed. The client will be responsible for calling `Status()` to get the final status.

File: supervisor/process.go (hunk 9 above)
Comment (alexmeli100) [reply supervisor/process.go]: Yeah you are definitely right that it would be simpler to just stream the bytes and let the client handle the output

File: supervisor/process.go (hunk 10 above)
Comment (alexmeli100) [reply supervisor/process.go]: The `Supervisor` holds a buffer to store prior logs for each process. The client can always call `GetOutput()` to get the logs at any time.

File: supervisor/process.go (hunk 8 above)
Comment (alexmeli100) [reply supervisor/process.go]: This is to handle carriage return in newlines but it might not be needed anymore since I will just stream the bytes

File: supervisor/process.go [hunk 12]
```diff
@@ -0,0 +1,408 @@
...
+	case Finished:
+		return "finished"
+	case Interrupted:
+		return "interrupted"
+	case Fatal:
+		return "fatal"
+	default:
+		return "unknown"
+	}
+}
```
Comment (Joerger) [inline supervisor/process.go]: IMO a custom string type makes more sense here, Go iota's are rarely the best way to define constants.
```suggestion
type ProcessState string

const (
	Initial ProcessState = "intial"
	Starting ProcessState = "starting"
	Running ProcessState = "running"
	Stopping ProcessState = "stopping"
	Finished ProcessState = "finished"
	Interrupted ProcessState = "interrupted"
	Fatal ProcessState = "failed"
)
```

File: supervisor/process.go [hunk 13]
```diff
@@ -0,0 +1,408 @@
...
+
+// Process is a wrapper around exec.Cmd to run linux processes
+type Process struct {
+	*sync.Mutex
+	cmd          *exec.Cmd
+	streamStdOut chan string
+	streamStdErr chan string
+	stdOut       *StreamOutput
+	stdErr       *StreamOutput
+	state        ProcessState
```
Comment (Joerger) [inline supervisor/process.go]: state is duplicated in status.

File: supervisor/process.go [hunk 14]
```diff
@@ -0,0 +1,408 @@
...
+func (p *Process) Start() <-chan Status {
+	p.Lock()
+	defer p.Unlock()
+
+	if p.statusChan != nil {
+		return p.statusChan
+	}
+
+	p.statusChan = make(chan Status, 1)
+	p.setState(Starting)
```
Comment (Joerger) [inline supervisor/process.go]: What is the purpose of the starting state? The process will always be either `Fatal` or `Running` at the end of the function. Since it's protected by a lock, the `Starting` state is never visible.

File: supervisor/process.go [hunk 15]
```diff
@@ -0,0 +1,408 @@
...
+
+	p.cmd.Stdout = p.stdOut
+	p.cmd.Stderr = p.stdErr
+
+	p.cmd.SysProcAttr = &syscall.SysProcAttr{Setpgid: true}
```
Comment (Joerger) [inline supervisor/process.go]: These can be set in the `NewProcess` constructor.

File: supervisor/process.go [hunk 16]
```diff
@@ -0,0 +1,408 @@
...
+		p.status.StopTime = time.Now().UnixNano()
+		p.status.Error = err
+		p.setState(Fatal)
+		p.statusChan <- p.status
+		close(p.done)
+
+		return p.statusChan
+	}
+
+	p.status.StartTime = now.UnixNano()
```
Comment (Joerger) [inline supervisor/process.go]:
```suggestion
	p.status.StartTime := time.Now().UnixNano()
	if err := p.cmd.Start(); err != nil {
		// process failed to start.
		// exit immediately
		p.status.StopTime = time.Now().UnixNano()
		p.status.Error = err
		p.setState(Fatal)
		p.statusChan <- p.status
		close(p.done)

		return p.statusChan
	}

```

File: supervisor/process.go (hunk 5 above)
Comment (Joerger) [reply supervisor/process.go]: What is the point of the `Interrupted` state then? Like `Starting`, it in immediately overwritten and never visible outside this function due to the lock.

File: supervisor/process.go [hunk 17]
```diff
@@ -0,0 +1,408 @@
...
+	return p.status
+}
+
+// Stop stops a process by sending it's group a SIGTERM signal
+func (p *Process) Stop(byUser bool) error {
+	p.Lock()
+	defer p.Unlock()
+
+	if p.IsInitialState() || p.IsFinalState() {
+		return nil
```
Comment (Joerger) [inline supervisor/process.go]: This should probably return an error like "Failed to stop process, already in Initial/Final state"

File: supervisor/process.go [hunk 18]
```diff
@@ -0,0 +1,408 @@
...
+	return syscall.Kill(-p.cmd.Process.Pid, syscall.SIGTERM)
+}
+
+// Signal sends an os signal to the process
+func (p *Process) Signal(sig syscall.Signal) error {
+	p.Lock()
+	defer p.Unlock()
+
+	if p.IsInitialState() || p.IsFinalState() {
+		return nil
```
Comment (Joerger) [inline supervisor/process.go]: This should probably return an error like "Failed to signal process, already in Initial/Final state"

File: supervisor/supervisor.go [hunk 19]
```diff
@@ -0,0 +1,311 @@
...
+	if cmd == "" {
+		msg := "cannot add job, no executable defined"
+		s.logger.Error(msg)
+		return nil, errors.New(msg)
+	}
+
+	_, err := exec.LookPath(cmd)
+
+	if err != nil {
+		msg := fmt.Sprintf("cannot add job, the system cannot find the given executable \"%s\"", cmd)
```
Comment (Joerger) [inline supervisor/supervisor.go]:
```suggestion
		msg := fmt.Sprintf("cannot start job, the system cannot find the given executable \"%s\"", cmd)
```

File: supervisor/supervisor.go [hunk 20]
```diff
@@ -0,0 +1,311 @@
...
+
+				watchers.sendLog(line)
+			case <-p.Done():
+				return
+			}
+
+		}
+	}(id, proc)
+
+	fs := <-sc
```
Comment (Joerger) [inline supervisor/supervisor.go]: Is this the main use of the status chan? Could you just receive from `p.Done()`, lock p.status, and then read the status?

File: design.md [hunk 21]
```diff
@@ -96,11 +96,14 @@ All TLS 1.3 cipher suites are supported
 * TLS_AES_256_GCM_SHA384
 * TLS_CHACHA20_POLY1305_SHA256
 * TLS_AES_128_GCM_SHA256
+* TLS_AES_128_CCM_8_SHA256
+* TLS_AES_128_CCM_SHA256
```
Comment (awly) [inline design.md:100]: `crypto/tls` in Go does not support these
```suggestion
```

File: design.md [hunk 22]
```diff
@@ -96,11 +96,14 @@ All TLS 1.3 cipher suites are supported
...
 ### Key generation algorithm
 Keys are generated using 4096 bit RSA
 
 ### Authorization
-When starting a job, the server assigns the job to the client using the common name of it's cerficate. A client can stop, get the status and get the output of a job it owns. Attempting to perform any of these operations for a job not owned by the client results in an `NOT_FOUND` error.
+When starting a job, the server assigns the job to the client using the common name of it's cerficate. A client can stop, get the status and get the output of a job it owns. Attempting to perform any of these operations for a job not owned by the client results in an `UNAUTHOURIZED` error.
```
Comment (awly) [inline design.md]: Looks like this was reverted from the original design doc
```suggestion
When starting a job, the server assigns the job to the client using the common name of it's cerficate. A client can stop, get the status and get the output of a job it owns. Attempting to perform any of these operations for a job not owned by the client results in an `NOT_FOUDN` error.
```

File: supervisor/process.go [hunk 23]
```diff
@@ -0,0 +1,408 @@
...
+
+// StdErr returns the standard error streaming channel
+func (p *Process) StdErr() chan string {
+	return p.streamStdErr
+}
+
+// Done returnes a channel that is closed when the process finishes running
+func (p *Process) Done() chan struct{} {
+	return p.done
+}
```
Comment (Joerger) [inline supervisor/process.go]: These channels should be returned as read only.

File: supervisor/process.go [hunk 24]
```diff
@@ -0,0 +1,408 @@
...
+	"github.com/pkg/errors"
+	"os/exec"
+	"strings"
+	"sync"
+	"syscall"
+	"time"
+)
+
+const (
+	DefaultLineBufferSize = 16384
```
Comment (awly) [inline supervisor/process.go]: curious: how did you choose this buffer size?

File: supervisor/process.go [hunk 25]
```diff
@@ -0,0 +1,408 @@
...
+type ProcessState uint8
+
+const (
+	Initial ProcessState = iota
+	Starting
+	Running
+	Stopping
+	Finished
+	Interrupted
+	Fatal
```
Comment (awly) [inline supervisor/process.go]: this looks different from https://github.com/alexmeli100/teleport-assessment/blob/master/api/grpc/service.proto#L52-L56
should the possible states be the same?

File: supervisor/process.go [hunk 26]
```diff
@@ -0,0 +1,408 @@
...
+	cmd          *exec.Cmd
+	streamStdOut chan string
+	streamStdErr chan string
+	stdOut       *StreamOutput
+	stdErr       *StreamOutput
+	state        ProcessState
+	status       Status
+	statusChan   chan Status
+	done         chan struct{}
+	stateLock    *sync.RWMutex
```
Comment (awly) [inline supervisor/process.go]: what's the difference between the top mutex and `stateLock`?
is it possible to only keep one mutex and avoid potential deadlocks?

File: supervisor/process.go [hunk 27]
```diff
@@ -0,0 +1,408 @@
...
+	case Fatal:
+		return "fatal"
+	default:
+		return "unknown"
+	}
+}
+
+// Process is a wrapper around exec.Cmd to run linux processes
+type Process struct {
+	*sync.Mutex
```
Comment (awly) [inline supervisor/process.go]: nit: don't embed mutexes.
this allows the external caller to mess with it and cause unexpected behavior

File: supervisor/process.go [hunk 28]
```diff
@@ -0,0 +1,408 @@
...
+// NewProcess creates and returns a new process without running it
+func NewProcess(name string, args []string) *Process {
+	a := strings.Join(args, " ")
+
+	status := Status{
+		Cmd:        name + " " + a,
+		PID:        0,
+		Error:      nil,
+		ExitCode:   -1,
+		StopByUser: false,
```
Comment (awly) [inline supervisor/process.go]: nit: you can skip the zero values
```suggestion
		Cmd:        name + " " + a,
		ExitCode:   -1,
```

File: supervisor/process.go [hunk 29]
```diff
@@ -0,0 +1,408 @@
...
+	return p.state == Running || p.state == Starting
+}
+
+// IsFinalState is true if process is Finished or Interrupted
+func (p *Process) IsFinalState() bool {
+	p.stateLock.RLock()
+	defer p.stateLock.RUnlock()
+
+	return p.state == Finished || p.state == Interrupted
+}
```
Comment (awly) [inline supervisor/process.go]: how about a `GetState` helper instead?

File: supervisor/process.go [hunk 30]
```diff
@@ -0,0 +1,408 @@
...
+
+// Start starts the process and returns a channel the caller can
+// use to retrieve the final status
+func (p *Process) Start() <-chan Status {
```
Comment (awly) [inline supervisor/process.go]: it looks like the return channel here is only used by the `Supervisor` to wait for all jobs to terminate.
you could simplify the implementation by having a `Wait` method on the `Process` instead

File: supervisor/process.go (hunk 14 above)
Comment (awly) [inline supervisor/process.go]: note: this state will never be seen
when this function returns, the state will always be changed to either Fatal or Running

File: supervisor/process.go [hunk 31]
```diff
@@ -0,0 +1,408 @@
...
+
+// Status is the status of a process
+type Status struct {
+	Cmd        string
+	PID        int
+	State      ProcessState
+	Error      error
+	ExitCode   int
+	StartTime  int64
+	StopTime   int64
```
Comment (awly) [inline supervisor/process.go]: can these fields be `time.Time`?

File: supervisor/process.go [hunk 32]
```diff
@@ -0,0 +1,408 @@
...
+
+	return p.statusChan
+}
+
+// wait waits for a process to finish running and sets it's final status
+func (p *Process) wait() {
+	// send the final status and close done channel to signal to
+	// goroutines the process has finished running
+	defer func() {
+		p.statusChan <- p.GetStatus()
```
Comment (awly) [inline supervisor/process.go]: this goroutine might leak if `statusChan` is not consumed anywhere

File: supervisor/process.go [hunk 33]
```diff
@@ -0,0 +1,408 @@
...
+	p.Lock()
+	defer p.Unlock()
+
+	// get the exit code of the process
+	if err != nil {
+		if exitErr, ok := err.(*exec.ExitError); ok {
+			err = nil
+
+			if status, ok := exitErr.Sys().(syscall.WaitStatus); ok {
+				exitCode = status.ExitStatus()
```
Comment (awly) [inline supervisor/process.go]:
```suggestion
			exitCode = exitErr.ExitCode()
			if status, ok := exitErr.Sys().(syscall.WaitStatus); ok {
```

File: supervisor/process.go (hunk 5 above)
Comment (awly) [reply supervisor/process.go]: is the Interrupted status needed then?

File: supervisor/process.go [hunk 34]
```diff
@@ -0,0 +1,408 @@
...
+}
+
+// Stop stops a process by sending it's group a SIGTERM signal
+func (p *Process) Stop(byUser bool) error {
+	p.Lock()
+	defer p.Unlock()
+
+	if p.IsInitialState() || p.IsFinalState() {
+		return nil
+	}
```
Comment (awly) [inline supervisor/process.go]:
```suggestion
	if p.state != Running {
		return nil
	}
```

File: supervisor/process.go [hunk 35]
```diff
@@ -0,0 +1,408 @@
...
+
+	p.setState(Stopping)
+	p.status.StopByUser = byUser
+
+	return syscall.Kill(-p.cmd.Process.Pid, syscall.SIGTERM)
```
Comment (awly) [inline supervisor/process.go]:
```suggestion
	return p.cmd.Process.Signal(syscall.SIGTERM)
```

File: supervisor/process.go [hunk 36]
```diff
@@ -0,0 +1,408 @@
...
+
+	if p.IsInitialState() || p.IsFinalState() {
+		return nil
+	}
+
+	if sig == syscall.SIGTERM {
+		p.setState(Stopping)
+	}
+
+	return syscall.Kill(-p.cmd.Process.Pid, sig)
```
Comment (awly) [inline supervisor/process.go]: No need to guess the state based on signal, it's not possible to predict what most signals will do
```suggestion
	if !p.state == Running {
		return nil
	}

	return p.cmd.Process.Signal(sig)
```

File: supervisor/process.go [hunk 37]
```diff
@@ -0,0 +1,408 @@
...
+	}
+
+	for _, opt := range opts {
+		opt(s)
+	}
+
+	return s
+}
+
+// TODO: stream bytes instead of lines
```
Comment (awly) [inline supervisor/process.go:208]: let's address this TODO now, it will massively simplify the code

**Review by awly — COMMENTED** [review]
Note: didn't get to `supervisor.go` or tests yet

File: supervisor/process.go [hunk 38]
```diff
@@ -0,0 +1,231 @@
...
+// Write implements the io.Writer interface for StreamOutput
+func (s *StreamOutput) Write(buf []byte) (int, error) {
+	n := len(buf)
+
+	for _, b := range buf {
+		s.streamChan <- b
+	}
```
Comment (awly) [inline supervisor/process.go]: making these channels be `chan []byte` would be much more efficient than sending 1 byte at a time

File: supervisor/supervisor.go [hunk 39]
```diff
@@ -0,0 +1,311 @@
...
+// is returned if the executable cannot be found by the system.
+// A unique id is generated for the job and returned in JobStatus
+func (s *Supervisor) Start(cmd string, args []string) (*JobStatus, error) {
+	if cmd == "" {
+		msg := "cannot start job, no executable defined"
+		s.logger.Error(msg)
+		return nil, errors.New(msg)
+	}
+
+	_, err := exec.LookPath(cmd)
```
Comment (awly) [inline supervisor/supervisor.go]: note: this would not be needed if `NewProcess` started the command right away (without a separate `Start` method) and returned the error from `p.cmd.Start` instead of using an internal `error` field

**Review by awly — APPROVED** [review]

File: supervisor/supervisor.go (hunk 39 above)
Comment (alexmeli100) [reply supervisor/supervisor.go]: Thanks. This has been fixed. The Process now starts right away in `NewProcess`

File: supervisor/process.go (hunk 38 above)
Comment (alexmeli100) [reply supervisor/process.go]: Thanks. This has been fixed
